subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/incremental/IncrementalNativeCompiler.java

```diff
@@ -69,7 +69,7 @@ public WorkResult execute(final T spec) {
         // For source files that do include the precompiled header, we mark them as a "source file for pch"
         // The native compiler then adds the appropriate compiler arguments for those source files that can use PCH
         if (spec.getPreCompiledHeader() != null) {
-            ImmutableList.Builder<File> sourceFiles = ImmutableList.builder();
+            ImmutableList.Builder<File> sourceFiles = ImmutableList.builderWithExpectedSize(spec.getSourceFiles().size());
```

Comment:
The source file is only added, if it uses the precompiled header. Do you think it is faster anyway to allocate a list of the size of all source files?

Reply:
I guess the goal with PCH is that you want to use them for headers that are used a lot, so I'd expect the source files that match to typically be close to all sources.

Using the builder has two effects:

- if you don't more entries than expected, you can avoid re-allocation during the "adding" phase,
- if you add exactly the expected number of elements, you can avoid copying the data at the end.

We probably don't hit the second optimization here, but if there are many source files with a significant number of them using the PCHs, then we avoid some rounds of re-allocations.

This is probably not the most significant optimization, though.

Reply:
The effects are what I expected, I just was not sure how close the allocated number vs the PCH number of sources would be. Your reasoning sounds fine, though we should keep this change in mind in case someone reports a regression Gradle 5.5 when using PCH. // cc @lacasseio